File: sierpinski_pentagon/sierpinski_pentagon.pl
main:-
    write_sierpinski_pentagon('sierpinski_pentagon.svg', 600, 5).

write_sierpinski_pentagon(File, Size, Order):-
    open(File, write, Stream),
    format(Stream, "<svg xmlns='http://www.w3.org/2000/svg' width='~d' height='~d'>\n",
       [Size, Size]),
    write(Stream, "<rect width='100%' height='100%' fill='white'/>\n"),
    Margin = 5,
    Radius is Size/2 - 2 * Margin,
    pi(Pi),
    Side is Radius * sin(Pi/5) * 2,
    X is Size/2,
    Y is 3 * Margin,
    Scale_factor is 1/(2 + cos(2 * Pi/5) * 2),
    sierpinski_pentagon(Stream, X, Y, Scale_factor, Side, Order),
    write(Stream, "</svg>\n"),
    close(Stream).

pi(3.14159265359).

sierpinski_pentagon(Stream, X, Y, _, Side, 1):-
    !,
    write(Stream, "<path stroke-width='1' stroke='black' fill='blue' d='"),
    format(Stream, 'M ~g,~g ', [X, Y]),
    pi(Pi),
    Angle is 6 * Pi/5,
    write_pentagon_points(Stream, Side, Angle, X, Y, 5),
    write(Stream, "z'/>\n").
sierpinski_pentagon(Stream, X, Y, Scale_factor, Side, N):-
    Side1 is Side * Scale_factor,
    N1 is N - 1,
    pi(Pi),
    Angle is 6 * Pi/5,
    sierpinski_pentagons(Stream, X, Y, Scale_factor, Side1, Angle, N1, 5).

write_pentagon_points(_, _, _, _, _, 0):-!.
write_pentagon_points(Stream, Side, Angle, X, Y, N):-
    N1 is N - 1,
    pi(Pi),
    X1 is X + cos(Angle) * Side,
    Y1 is Y - sin(Angle) * Side,
    Angle1 is Angle + 2 * Pi/5,
    format(Stream, 'L ~g,~g ', [X1, Y1]),
    write_pentagon_points(Stream, Side, Angle1, X1, Y1, N1).

sierpinski_pentagons(_, _, _, _, _, _, _, 0):-!.
sierpinski_pentagons(Stream, X, Y, Scale_factor, Side, Angle, N, I):-
    I1 is I - 1,
    pi(Pi),
    Distance is Side + Side * cos(2 * Pi/5) * 2,
    X1 is X + cos(Angle) * Distance,
    Y1 is Y - sin(Angle) * Distance,
    Angle1 is Angle + 2 * Pi/5,
    sierpinski_pentagon(Stream, X1, Y1, Scale_factor, Side, N),
    sierpinski_pentagons(Stream, X1, Y1, Scale_factor, Side, Angle1, N, I1).

File: sierpinski_pentagon/sierpinski_pentagon.rs
// [dependencies]
// svg = "0.8.0"

fn sierpinski_pentagon(mut document : svg::Document, mut x : f64, mut y : f64,
        mut side : f64, order: usize) -> svg::Document {
    use svg::node::element::Path;
    use svg::node::element::path::Data;
    use std::f64::consts::PI;

    let degrees72 = 0.4 * PI;
    let mut angle = 3.0 * degrees72;
    let scale_factor = 1.0/(2.0 + degrees72.cos() * 2.0);

    if order == 1 {
        let mut data = Data::new();
        data = data.move_to((x, y));
        for _ in 0..5 {
            x += angle.cos() * side;
            y -= angle.sin() * side;
            angle += degrees72;
            data = data.line_to((x, y));
        }
        data = data.close();
        let path = Path::new()
            .set("fill", "blue")
            .set("stroke", "black")
            .set("stroke-width", "1")
            .set("d", data);
        document = document.add(path);
    } else {
        side *= scale_factor;
        let distance = side + side * degrees72.cos() * 2.0;
        for _ in 0..5 {
            x += angle.cos() * distance;
            y -= angle.sin() * distance;
            angle += degrees72;
            document = sierpinski_pentagon(document, x, y, side, order - 1);
        }
    }
    document
}

fn write_sierpinski_pentagon(file : &str, size : usize,
        order : usize) -> std::io::Result<()> {
    use svg::node::element::Rectangle;
    use std::f64::consts::PI;

    let margin = 5.0;
    let radius = (size as f64)/2.0 - 2.0 * margin;
    let side = radius * (0.2 * PI).sin() * 2.0;
    let x = (size as f64)/2.0;
    let y = 3.0 * margin;

    let rect = Rectangle::new()
        .set("width", "100%")
        .set("height", "100%")
        .set("fill", "white");

    let mut document = svg::Document::new()
        .set("width", size)
        .set("height", size)
        .add(rect);

    document = sierpinski_pentagon(document, x, y, side, order);
    svg::save(file, &document)
}

fn main() {
    write_sierpinski_pentagon("sierpinski_pentagon.svg", 600, 5).unwrap();
}